### PR TITLE
Move lifting to distance evaluation

### PR DESCRIPTION
**Pull request by iliailia**
After introduction of gRPC networking, our current runtime bottleneck is contained in the less than function that calls the `cross_compare` routine. It takes roughly 70% of total time in search/insert benchmarks. The number of `cross_compare` calls is O(k*n) where k is the number of neighbors observed in each layer per inserted value and n is the graph size.

Our current implementation of cross_compare does:
- lift (7 rounds)
  - conversion to bits: 5 rounds
  - bit injection: 2 rounds (can be done with one with more message sent)
- multiplication of lifted values (1 round)
- MSB extraction out of 32-bit value (6 rounds)

So in total 14 rounds per each call, or ~14kn in total to insert n values.

On the other hand, the number of `eval_distance` calls is linear only in n. So it makes sense to move as much preprocessing for `cross_compare` to distance evaluation. 

This PR implements a low-hanging fruit of moving lifting from u16 to u32 shares to `eval_distance`. This effectively reduces the cost of `cross_compare` by half to 7 rounds.

Benchmark results:
```bash
gr_ready_made_hnsw/gr-big-hnsw-insertions/1
                        time:   [4.4274 ms 4.4916 ms 4.5383 ms]
                        change: [+14.699% +23.091% +30.221%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-searches/1
                        time:   [5.9869 ms 6.0510 ms 6.1412 ms]
                        change: [-7.0358% -5.0068% -2.9948%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-insertions/10
                        time:   [152.45 ms 155.57 ms 161.13 ms]
                        change: [-35.514% -33.474% -31.385%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-searches/10
                        time:   [99.279 ms 101.32 ms 103.30 ms]
                        change: [-27.817% -25.752% -23.687%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-insertions/100
                        time:   [2.1942 s 2.2430 s 2.2976 s]
                        change: [-41.062% -39.454% -37.584%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-searches/100
                        time:   [1.8666 s 1.9233 s 1.9840 s]
                        change: [-37.998% -36.073% -33.714%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-insertions/1000
                        time:   [5.3318 s 5.4321 s 5.5099 s]
                        change: [-39.185% -37.802% -36.627%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-searches/1000
                        time:   [5.0458 s 5.1441 s 5.2482 s]
                        change: [-37.465% -36.117% -34.928%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-insertions/10000
                        time:   [11.950 s 12.108 s 12.254 s]
                        change: [-38.239% -37.298% -36.519%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-searches/10000
                        time:   [11.603 s 11.800 s 11.981 s]
                        change: [-39.749% -38.647% -37.513%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-insertions/100000
                        time:   [20.175 s 20.338 s 20.523 s]
                        change: [-38.044% -37.473% -36.928%] (p = 0.00 < 0.05)
gr_ready_made_hnsw/gr-big-hnsw-searches/100000
                        time:   [19.938 s 20.083 s 20.230 s]
                        change: [-37.458% -36.996% -36.493%] (p = 0.00 < 0.05)
```

TODO:
Precompute `code_dot/mask_dot` fractions using Goldshmidt division such that the resulting distances are u16 shares. In this case, the preprocessing stage contains
- lifting to u32 (7 rounds, see below),
- Make the initial approximation (usually `1/2^(k+1)` where `2^k < denominator <= 2^(k+1)`). 
  - bit conversion of 32-bit denominator (6 rounds).
  - Compute `1/2^(k+1)` as a fixed point value (6 rounds).
- Then we do Goldschmidt division since it needs only 2 concurrent fixed-point multiplications per iteration unlike the Newton-Raphson method. 4 iterations are enough to get results with a 16-bit precision. Each fixed-point multiplication can be done in 1, 2 or 3 rounds depending on the type of truncation algorithm. So we need between 4 and 12 rounds for this step.

So we need 23-31 rounds per each distance in the preprocessing stage. So if we insert n values, it’ll total 23n-31n rounds.
In the active phase, we do comparisons containing now one subtraction and a single MSB extraction, which is 5 rounds per 16-bit value.

### CODE REVIEW

**Review by dkales — APPROVED** [review]
LGTM

**Review by dkales — APPROVED** [review]

**Review by dkales — APPROVED** [review]

File: iris-mpc-cpu/src/protocol/ops.rs [hunk 1]
```diff
@@ -263,6 +250,8 @@ pub async fn galois_ring_is_match(
     assert_eq!(pairs.len(), 1);
     let additive_dots = galois_ring_pairwise_distance(session, pairs).await?;
     let rep_dots = galois_ring_to_rep3(session, additive_dots).await?;
+    let rep_dots = VecShare::new_vec(rep_dots);
+    let rep_dots = batch_signed_lift(session, rep_dots).await?.inner();
     // compute dots[0] - dots[1]
     let bit = compare_threshold(session, rep_dots[0].clone(), rep_dots[1].clone()).await?;
```
[the remark this answers is not in the document]
Comment (bgillesp) [reply iris-mpc-cpu/src/protocol/ops.rs:256]: (Also I think that `batch_signed_lift` can be translated to `batch_signed_lift_vec` here without the surrounding boilerplate.)

[the remark this answers is not in the document]
Comment (bgillesp) [reply iris-mpc-cpu/src/protocol/ops.rs:256]: Possible approach to preserve the lifting optimization: have two functions, `compare_threshold` as written which uses pre-lifted 32-bit shares, and `lift_and_compare_threshold` which acts like the previous version taking 16-bit shares as input and uses the optimized lifting and calls, and calls `compare_threshold` after lifting.  The first can be used in `compare_threshold_and_open`, and the second in `galois_ring_is_match`.

**Review by bgillesp — APPROVED** [review]
Looking good overall!  A few small suggestions, otherwise should be good to merge.

File: iris-mpc-cpu/src/protocol/ops.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (iliailia) [reply iris-mpc-cpu/src/protocol/ops.rs]: Thanks. Completely missed this optimization. The functions are adjusted as you suggested.

**Pull request merged**